File: 2018/09_MarbleMania/game.py
# ======================================================================
# Marble Mania
#   Advent of Code 2018 Day 09 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                            g a m e . p y
# ======================================================================
"A solver for the Advent of Code 2018 Day 09 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
from collections import deque

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
CLOCKWISE = -1
COUNTER_CLOCKWISE = 1

# ======================================================================
#                                                                   Game
# ======================================================================


class Game(object):   # pylint: disable=R0902, R0205
    "Object for Marble Mania"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.num_players = 0
        self.last_marble = 0
        self.current_marble = 0
        self.current_player = None
        self.circle = deque([0])
        self.players = []

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            words = text[0].split()
            self.num_players = int(words[0])
            self.last_marble = int(words[6])
            self.players = [0] * self.num_players

    def highest_score(self):
        if len(self.players) > 0:
            return max(self.players)
        else:
            return 0

    def play_game(self):
        while self.game_turn() is not None:
            pass
        return self.highest_score()

    def game_turn(self):
        marble = self.next_marble()
        if marble is None:
            return None
        self.next_player()
        if marble % 23 != 0:
            self.place_marble()
        else:
            self.capture_marble()
        return marble

    def next_marble(self):
        if self.current_marble >= self.last_marble:
            return None
        self.current_marble += 1;
        return self.current_marble

    def next_player(self):
        if self.current_player is None:
            self.current_player = 0
        else:
            self.current_player = (self.current_player + 1) % self.num_players
        return self.current_player

    def place_marble(self):
        self.circle.rotate(CLOCKWISE)
        self.circle.append(self.current_marble)

    def capture_marble(self):
        # 1. Player gets to keep marble
        #print("capturing player = %d" % self.current_player)
        self.players[self.current_player] += self.current_marble
        # 2. In addition, the marble 7 marbles counter-clockwise from the current marble
        #    is removed from the circle and also added to the current player's score.
        #    The marble located immediately clockwise of the marble that was removed
        #    becomes the new current marble.
        self.circle.rotate(7 * COUNTER_CLOCKWISE)
        self.players[self.current_player] += self.circle.pop()
        self.circle.rotate(CLOCKWISE)

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 1. Return the solution for part one
        return self.play_game()


    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 1. Return the solution for part two
        self.last_marble *= 100
        return self.play_game()

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                           g a m e . p y                        end
# ======================================================================

File: 2018/09_MarbleMania/game.go
// ======================================================================
// Marble Mania
//   Advent of Code 2018 Day 09 -- Eric Wastl -- https://adventofcode.com
//
// lua implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         g a m e . g o
// ======================================================================
// Game for the Advent of Code 2018 Day 09 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"strings"
)

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const ()

// ======================================================================
//                                                                   Game
// ======================================================================

type Game struct {
	numPlayers int
	numMarbles int
	players    []int
	circle     []int
	nextPlayer int
	nextMarble int
	current    int
}

func NewGame(players int, marbles int) *Game {
	// Game Object for Marble Mania

	// 1. Set the initial values
	p := new(Game)
	p.numPlayers = players
	p.numMarbles = marbles
	p.nextPlayer = 0
	p.nextMarble = 1
	p.current = 0
	p.players = make([]int, p.numPlayers)
	p.circle = make([]int, 1, p.numMarbles)

	// 2. Return Game
	return p
}

func (p *Game) Show() string {
	// Display the game

	// 1. Kludge for the start-+
	if p.nextPlayer == 0 {
		return "[----] (0)"
	}

	// 2. Start with nothing
	var result strings.Builder

	// 3. Add the player
	fmt.Fprintf(&result, "[%3d] ", p.nextPlayer)

	// 4. Loop for all the marbles in the circle
	for indx, marble := range p.circle {

		// 5. if this is the current marble use () else spaces
		if indx == p.current {
			fmt.Fprintf(&result, "(%d)", marble)
		} else {
			fmt.Fprintf(&result, " %d ", marble)
		}
	}

	// 6. Return the game information as a string
	return result.String()
}
func (p *Game) Turn() bool {
	// Do a turn, return true if the game should continue

	// 1. Get the marble
	marble := p.nextMarble
	if marble > p.numMarbles {
		return false
	}
	p.nextMarble = marble + 1

	// 2. Get the player
	player := p.nextPlayer
	p.nextPlayer = player + 1
	if p.nextPlayer >= p.numPlayers {
		p.nextPlayer = 0
	}

	// 3. Is this a special or regular turn?
	if p.IsSpecial(marble) {
		p.SpecialTurn(marble, player)
	} else {
		p.RegularTurn(marble, player)
	}

	// 4. Go again
	return true
}

func (p *Game) AllTurns() int {
	// Play the game until the end

	// 1. Lots of turns
	for p.Turn() {
	}

	// 2. Return the highest score
	return p.Highest()
}

func (p *Game) Highest() int {
	// Return the highest game score

	// 1. Start with nothing
	highest := 0

	// 2. Loop for all the players
	for _, score := range p.players {

		// 3. Keep the highest score
		if score > highest {
			highest = score
		}
	}

	// 4. Return the highest score
	return highest
}

func (p *Game) RegularTurn(marble int, player int) {
}

func (p *Game) SpecialTurn(marble int, player int) {
}

func (p *Game) Insert(index, value int) {
	// Insert a marble into the circle between the index and index + 1

	// 1. Adding to the end is easy
	if index > len(p.circle) {
		p.circle = append(p.circle, value)
	}

}

func (p *Game) Remove(slice []int, index int) int {
	// Remove a marble from the circle at the index value

	marble := 0

	// 9. Return removed marble
	return marble
}

func (p *Game) Clockwise() int {
	// Returns the index of the marble one space clockwise of the current one

	// 1. Clockwise is an increase in the index
	next := p.current + 1

	// 2. If greater than the current lenght, circle around
	if next > len(p.circle) {
		next = next - len(p.circle)
	}

	// 3. Return the index
	return next
}

func (p *Game) CounterClockwise() int {
	// Returns the index of the marble seven spaces cc of the current one

	// 1. Counter Clockwise is an decrease in the index
	next := p.current - 7

	// 2. If less than the start of the array, circle around
	if next < 0 {
		next = len(p.circle) + next
	}

	// 3. Return the index
	return next
}

func (p *Game) IsSpecial(marble int) bool {
	// Returns true if this is a special marble

	return 0 == marble%23
}

// ======================================================================
// end                          g a m e . g o                         end
// ======================================================================
